alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,263 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import namedtuple
+
+import gin.tf
+import tensorflow as tf
+
+from tf_agents.networks.network import Network
+import tf_agents.specs.tensor_spec as tensor_spec
+
+from alf.algorithms.algorithm import Algorithm, AlgorithmStep, LossInfo
+from alf.utils.data_buffer import DataBuffer
+from tf_agents.trajectories.time_step import StepType
+from alf.utils.conditional_ops import conditional_update, run_if
+
+MISCInfo = namedtuple("MISCInfo", ["reward", "loss"])
+
+
+@gin.configurable
+class MISCAlgorithm(Algorithm):
+    """Mutual Information-based State Control (MISC)
+    Author: Rui Zhao 
+    Work done during a research internship at Horizon Robotics. 
+    The paper is currently under review in a conference.
+
+    This algorithm generates the intrinsic reward based on the mutual information estimation between the states of interests and the context states.
+
+    See Zhao et al "Self-Supervised State-Control through Intrinsic Mutual Information Rewards"
+    """
+
+    def __init__(self,
+                 feature_spec,
+                 traj_spec,
+                 soi_spec,
+                 soc_spec,
+                 mi_r_scale=5000,
+                 hidden_size=256,
+                 misc_layerx: Network = None,
+                 misc_layery: Network = None,
+                 misc_layero: Network = None,
+                 buffer_size=100,
+                 n_objects=1,
+                 empowerment=False,
+                 env_name='SocialBot-PlayGround-v0',
+                 name="MISCAlgorithm"):
+        """Create an MISCAlgorithm.
+        
+        Args:
+            feature_spec: size of the state and the action
+            traj_spec: (num_parallel_environments, feature_spec)
```

Comment:
Follow the style for argument comment:
 arg_name (type): explanation

Reply:
added